sdks/java/io/kinesis/src/test/java/org/apache/beam/sdk/io/kinesis/KinesisMockWriteTest.java

```diff
@@ -0,0 +1,278 @@
+  public void beforeTest() {
+    KinesisServiceMock kinesisService = KinesisServiceMock.getInstance();
+    kinesisService.init(STREAM, 1);
+  }
+
+  @Test
+  public void testWriteBuildsCorrectly() {
+    Properties properties = new Properties();
+    properties.setProperty("KinesisEndpoint", "localhost");
+    properties.setProperty("KinesisPort", "4567");
+
+    KinesisIO.Write write =
+        KinesisIO.write()
+            .withStreamName(STREAM)
+            .withPartitionKey(PARTITION_KEY)
+            .withPartitioner(new BasicKinesisPartitioner())
+            .withAWSClientsProvider(new FakeKinesisProvider())
+            .withProducerProperties(properties)
+            .withAttempts(10);
+
+    assertEquals(STREAM, write.getStreamName());
+    assertEquals(PARTITION_KEY, write.getPartitionKey());
+    assertEquals(properties, write.getProducerProperties());
+    assertEquals(FakeKinesisProvider.class, write.getAWSClientsProvider().getClass());
+    assertEquals(BasicKinesisPartitioner.class, write.getPartitioner().getClass());
+    assertEquals(10, write.getAttempts());
+
+    assertEquals("localhost", write.getProducerProperties().getProperty("KinesisEndpoint"));
+    assertEquals("4567", write.getProducerProperties().getProperty("KinesisPort"));
+  }
+
+  @Test
+  public void testWriteValidationFailsMissingStreamName() {
+    KinesisIO.Write write =
+        KinesisIO.write()
+            .withPartitionKey(PARTITION_KEY)
+            .withAWSClientsProvider(new FakeKinesisProvider());
+
+    thrown.expect(IllegalArgumentException.class);
+    write.expand(null);
+  }
+
+  @Test
+  public void testWriteValidationFailsMissingPartitioner() {
+    KinesisIO.Write write =
+        KinesisIO.write().withStreamName(STREAM).withAWSClientsProvider(new FakeKinesisProvider());
+
+    thrown.expect(IllegalArgumentException.class);
+    write.expand(null);
+  }
+
+  @Test
+  public void testWriteValidationFailsPartitionerAndPartitioneKey() {
+    KinesisIO.Write write =
+        KinesisIO.write()
+            .withStreamName(STREAM)
+            .withPartitionKey(PARTITION_KEY)
+            .withPartitioner(new BasicKinesisPartitioner())
+            .withAWSClientsProvider(new FakeKinesisProvider());
+
+    thrown.expect(IllegalArgumentException.class);
+    write.expand(null);
+  }
+
+  @Test
+  public void testWriteValidationFailsMissingAWSClientsProvider() {
+    KinesisIO.Write write =
+        KinesisIO.write().withPartitionKey(PARTITION_KEY).withStreamName(STREAM);
+
+    thrown.expect(IllegalArgumentException.class);
+    write.expand(null);
+  }
+
+  @Test
+  public void testNotExistedStream() {
+    KinesisIO.Write write =
+        KinesisIO.write()
+            .withStreamName(STREAM)
+            .withPartitionKey(PARTITION_KEY)
+            .withAWSClientsProvider(
+                new FakeKinesisProvider(getMockedAmazonKinesisClient(404)));
+
+    thrown.expect(IllegalArgumentException.class);
+    write.expand(null);
+  }
+
```

Comment:
beware, any validation that requires a call to the backend must be done in the IO only at runtime in `@setup` for your writeFn (not at pipeline construction time) because sometimes the backend is not accessible before the actual run of the pipeline.

Reply:
yes too early is what I meant. I've seen cases (see https://issues.apache.org/jira/browse/BEAM-2541) where users create the graph on their dev machines that do not have access to the backend and then deploy and run pipeline in dataflow. `expand()` is invoked at construction time, you need to do the check at runtime. I had this discussion with @jkff as well on ES5 PR